Add application instance deregistration logic

rtorrero opened the pull request:
# Description

This PR adds the application instance deregistration logic to the sap_system aggregate. Upon deregistering a host, it should remove the relevant application instances and eventually sap systems, depending on the role of the deregistered host.

Note: This PR derives from https://github.com/trento-project/web/pull/1343 and is a subset of it. A follow-up PR will contain the database instance deregistration logic.

## How was this tested?
Based on the deregistration flow chart, all of the possible scenarios have been added to the test cases.

fabriziosestito commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -290,6 +294,72 @@ defmodule Trento.Domain.SapSystem do
+  end
+
+  def apply(
+        %SapSystem{application: %Application{instances: []}} = sap_system,
```
no need for this match, just match `%SapSystem{} = sap_system`

fabriziosestito commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -290,6 +294,72 @@ defmodule Trento.Domain.SapSystem do
+  def apply(
+        %SapSystem{application: %Application{instances: []}} = sap_system,
+        %SapSystemDeregistered{
+          sap_system_id: sap_system_id,
```
no need for this

fabriziosestito commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -290,6 +294,72 @@ defmodule Trento.Domain.SapSystem do
+      ) do
+    %SapSystem{
+      sap_system
+      | sap_system_id: sap_system_id,
```
why update the sap_system_id here? no need for this

fabriziosestito commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -633,6 +703,36 @@ defmodule Trento.Domain.SapSystem do
+       end) ||
+         !Enum.any?(instances, fn %Instance{features: features} ->
+           String.contains?(features, "ABAP")
+         end) do
```
Maybe a bit cleaner with the help of variables, `unless` and text-match operator:

```
    has_abap? = Enum.any?(instances, fn instance -> instance.features =~ "ABAP" end)

    has_messageserver? =
      Enum.any?(instances, fn instance -> instance.features =~ "MESSAGESERVER" end)

    unless has_abap? && has_messageserver? do
    ....
```

arbulu89 replied:
Other solution could be:
```
# We map all the available features
instances
|> Enum.map(fn %{features: features} -> features end)
|> Enum.uniq()

# Check if message server and abap ara available in our list checking if the subset is there (https://stackoverflow.com/questions/63143363/elixir-check-array-contains-all-the-values-of-another-array)
unless Enum.empty?(["MESSAGESERVER", "ABAP"] -- features) do
   ...
end
```

fabriziosestito commented:
@rtorrero @jamie-suse @arbulu89 I sent some preliminary comments as I will be off till Wednesday. It is a very superficial review, so feel free to have a deeper discussion next week.

fabriziosestito commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -290,6 +294,72 @@ defmodule Trento.Domain.SapSystem do
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_sap_system_deregistration_events(
+        sap_system,
+        command
```
why command here? we just need deregistered_at

arbulu89 reviewed (commented):
Looks great @rtorrero 

I would still work in the `maybe_..` function. Both solutions proposed by me and Fabrizio look easier to understand in my opinion

arbulu89 commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -633,6 +701,33 @@ defmodule Trento.Domain.SapSystem do
+         },
+         deregistered_at
+       ) do
+    has_abap? = Enum.any?(instances, fn instance -> instance.features =~ "ABAP" end)
```
Maybe `fn %{features: features} -> ` better

arbulu89 reviewed (approved):
Alright @rtorrero 
It looks good in any case

arbulu89 commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -633,6 +701,33 @@ defmodule Trento.Domain.SapSystem do
+          deregistered_at: deregistered_at
+        }
+      end) ++
+        [%SapSystemDeregistered{sap_system_id: sap_system_id, deregistered_at: deregistered_at}]
```
Hey @rtorrero 
I have noticed that this event should be conditional.
We cannot send this event always as the sap system might not even be registered.
Example:
- We register only message server
- We deregister this same host
At this point, we shouldn't be sending the sap system deregistered event

rtorrero replied:
I've added a test for this particular scenario and fixed the code to handle this situation properly. Thanks!

arbulu89 reviewed (changes requested):
Hey @rtorrero @jamie-suse 
I have realized that we might have some issue. Could you check my last comment below?
PD: Besides that, i'm having doubts yet again with the operation of deregistering all the instances at once

rtorrero commented:
@arbulu89 Thanks for the comments, I've:
 - Applied the additional style improvements you suggested
 - Updated the aggregate logic to not deregister all instances when the last ABAP / messageserver instance gets deregistered
 - Updated the logic to only emit SapSystemDeregistered if it was previously registered. Note that this is currently causing the test to fail as we need pr#1358 from @CDimonaco  to be merged first
 
 As the last point indicates, we need another PR to land first so I'm setting this one to draft meanwhile.

rtorrero commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -701,30 +701,23 @@ defmodule Trento.Domain.SapSystem do
-        }
-      end) ++
-        [%SapSystemDeregistered{sap_system_id: sap_system_id, deregistered_at: deregistered_at}]
+    unless has_abap? and has_messageserver? and sid != nil do
```
I used `!is_nil(sid)` which I think is more appropriate here, thanks!

arbulu89 replied:
https://github.com/trento-project/web/pull/1357#discussion_r1176564351

arbulu89 commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -633,6 +701,26 @@ defmodule Trento.Domain.SapSystem do
+           }
+         },
+         deregistered_at
+       ) do
```
Come on @rtorrero ,
Don't fall in the old habits already :smiley_cat: 
```
defp maybe_emit_sap_system_deregistered_event(%SapSystem{sid: nil}), do: []
defp maybe_emit_sap_system_deregistered_event(
         %SapSystem{
...
#what you had before
```

arbulu89 commented on lib/trento/domain/sap_system/sap_system.ex:
```diff
@@ -290,6 +294,70 @@ defmodule Trento.Domain.SapSystem do
+      ) do
+    %SapSystem{
+      sap_system
+      | deregistered_at: deregistered_at
```
Maybe we should put `sid: nil` here. 
Otherwise, this instance cannot be de-registered never again.
Anyway, let's simply keep an eye on this for the restoration work

arbulu89 commented on test/trento/domain/sap_system/sap_system_test.exs:
```diff
@@ -1064,4 +1073,466 @@ defmodule Trento.SapSystemTest do
+      )
+    end
+
+    test "should deregister an ABAP Application Instance" do
```
maybe add `without deregistering the SAP system`

arbulu89 commented on test/trento/domain/sap_system/sap_system_test.exs:
```diff
@@ -1064,4 +1073,466 @@ defmodule Trento.SapSystemTest do
+      )
+    end
+
+    test "should deregister just a Message Server from a not-fully-registered SAP system" do
```
I would remove this test, it does exactly the same than the first. There is no any distinction in reality in the code (we know that elixis'r pattern matching works)

jamie-suse replied:
We checked this code path/scenario for coverage using _Coveralls_; this test covers the scenario where a SAP System is **not** registered, and _then_ an instance is deregistered - testing that a `SapSystemDeregistered` event is **not** emitted in this case. This is not the same as the first test - removing an ENQREP server, as it would never emit a `SapSystemDeregistered` event.

arbulu89 reviewed (approved):
Thanks @rtorrero 
Good to go from my side. Added some minor comments about the tests
